sdk/core/System.ClientModel/src/Convenience/ClientCache.cs

```diff
@@ -0,0 +1,148 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+
+namespace System.ClientModel.Primitives;
+
+/// <summary>
+/// A cache for storing client instances with a Least Recently Used (LRU) eviction policy.
+/// </summary>
+public class ClientCache
+{
+    private readonly int _maxClients = 100;
+    private readonly ConcurrentDictionary<(Type Type, string Id), LinkedListNode<CacheItem>> _cacheMap = new();
+    // Doubly linked list to maintain LRU order. The head is the most recently used, the tail is the least.
+    private readonly LinkedList<CacheItem> _lruList = new();
+    // ReaderWriterLockSlim to allow concurrent reads while synchronizing writes.
+    private readonly ReaderWriterLockSlim _lock = new(LockRecursionPolicy.NoRecursion);
+
+    /// <summary>
+    /// Retrieves a client from the cache or creates a new one if it does not exist.
+    /// </summary>
+    /// <typeparam name="T">The type of the client.</typeparam>
+    /// <param name="createClient">A factory function to create the client if not cached.</param>
+    /// <param name="id">An optional identifier for the client instance.</param>
+    /// <returns>The cached or newly created client instance.</returns>
+    public T GetClient<T>(Func<T> createClient, string id = "") where T : class
+    {
+        (Type Type, string Id) clientKey = (typeof(T), id ?? string.Empty);
+
+        // Try to retrieve an existing client.
+        _lock.EnterUpgradeableReadLock();
+        try
+        {
+            if (_cacheMap.TryGetValue(clientKey, out var node))
+            {
+                // Update the LRU list: move the accessed node to the front.
+                _lock.EnterWriteLock();
+                try
+                {
+                    _lruList.Remove(node);
+                    node.Value.LastAccessTime = DateTime.UtcNow;
```

Comment:
This is an expensive operation. https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.stopwatch.gettimestamp?view=net-9.0 is significantly cheaper.